crates/ntx-builder/src/state/mod.rs

```diff
@@ -0,0 +1,278 @@
+                for tx in txs {
+                    self.revert_transaction(tx);
+                }
+            },
+        }
+    }
+
+    /// Handles a [`MempoolEvent::TransactionAdded`] event.
+    fn add_transaction(
+        &mut self,
+        id: TransactionId,
+        nullifiers: Vec<Nullifier>,
+        network_notes: Vec<NetworkNote>,
+        account_delta: Option<AccountUpdateDetails>,
+    ) {
+        // Skip transactions we already know about.
+        //
+        // This can occur since both ntx builder and the mempool might
+        // inform us of the same transaction. Once when it was submitted to
+        // the mempool, and once by the mempool event.
+        if self.inflight_txs.contains_key(&id) {
+            return;
+        }
+
+        let mut tx_impact = Impact::default();
+        if let Some(update) = account_delta.and_then(NetworkAccountUpdate::from_protocol) {
+            tx_impact.account_delta = Some(update.prefix());
+            self.account_or_default(update.prefix()).add_delta(update);
+        }
+        for note in network_notes {
+            tx_impact.notes.insert(note.nullifier());
+            self.nullifier_idx.insert(note.nullifier(), note.account_prefix());
+            self.account_or_default(note.account_prefix()).add_note(note);
+        }
+        for nullifier in nullifiers {
+            // Ignore nullifiers that aren't network note nullifiers.
+            let Some(account) = self.nullifier_idx.get(&nullifier) else {
+                continue;
+            };
+            tx_impact.nullifiers.insert(nullifier);
+            // We don't use the entry wrapper here because the account must already exist.
+            self.accounts
+                .get_mut(account)
+                .expect("nullifier account must exist")
+                .add_nullifier(nullifier);
+        }
+
+        if !tx_impact.is_empty() {
+            self.inflight_txs.insert(id, tx_impact);
+        }
+    }
+
+    /// Grants mutable access to the given account state, creating a default entry if none exists.
+    ///
+    /// This is effectively a thin wrapper around the entry API, but this also tracks new
+    /// accounts to the queue.
+    ///
+    /// This _must_ be the only way new accounts are added as otherwise they won't be queued.
+    fn account_or_default(&mut self, prefix: NetworkAccountPrefix) -> &mut AccountState {
+        match self.accounts.entry(prefix) {
+            Entry::Occupied(occupied_entry) => occupied_entry.into_mut(),
+            Entry::Vacant(vacant_entry) => {
+                self.queue.push_back(prefix);
+                vacant_entry.insert(AccountState::default())
+            },
+        }
+    }
+
+    /// Handles [`MempoolEvent::BlockCommitted`] events.
+    fn commit_transaction(&mut self, tx: TransactionId) {
+        let Impact { account_delta, notes: _, nullifiers } =
+            self.inflight_txs.remove(&tx).unwrap_or_default();
```

Comment:
Same question as above: under what conditions would we not find an inflight tx for a committed transaction?